### PR TITLE
Pass WP_REST_Request to get_allowed_query_vars and containing filters…

### LINKED ISSUE #2821: Pass request to rest_query_vars filter
I want to be able to allow additional query parameters only when a specific request parameter exists to perform a meta query and ultimately return some posts.

Currently I've created a completely new endpoint that creates a new instance of `WP_REST_Posts_Controller` and ultimately returns the same output as the posts endpoint. 

I propose that we pass the `WP_REST_Request` instance through the constructor of `WP_REST_Posts_Controller::get_allowed_query_vars` and then pass it as a parameter in the `rest_query_vars` filter.

I'm quite sure other people might want to allow different query parameters for different requests but keep them restricted on others.

What do you think about this?

https://github.com/WP-API/WP-API/blob/develop/lib/endpoints/class-wp-rest-posts-controller.php#L665

https://github.com/WP-API/WP-API/blob/develop/lib/endpoints/class-wp-rest-posts-controller.php#L757

### PR DESCRIPTION
**Pull request by tobeycodes**
Resolves #2821

### CODE REVIEW

File: lib/endpoints/class-wp-rest-posts-controller.php
Comment (rmccue) [inline lib/endpoints/class-wp-rest-posts-controller.php]: This is a [hook in core](https://developer.wordpress.org/reference/hooks/query_vars/) which we're duplicating here for technical reasons, so I'm not sure we _can_ pass the request here. Is `rest_query_vars` not powerful enough for this?

**Review by rmccue — CHANGES_REQUESTED** [review]

**Comment by schrapel** [comment]
It is. I just want to cover all grounds. I'll update

File: lib/endpoints/class-wp-rest-posts-controller.php
Comment (tobeycodes) [reply lib/endpoints/class-wp-rest-posts-controller.php]: It is fine. Wanted to cover all grounds. I've updated the PR

**Comment by kadamwhite** [comment]
This is waiting on @rmccue's final review, but with the changes I'm +1 on this. Thanks for the patch @schrapel !

**Review by rmccue — APPROVED** [review]
Thanks a bunch! :)

**Pull request merged**
